// @naturalcycles/nodejs-lib/dist/colors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.boldCyan = exports.dimCyan = exports.cyan = exports.boldMagenta = exports.dimMagenta = exports.magenta = exports.boldBlue = exports.dimBlue = exports.blue = exports.boldRed = exports.dimRed = exports.red = exports.boldGreen = exports.dimGreen = exports.green = exports.inverseYellow = exports.boldYellow = exports.dimYellow = exports.yellow = exports.boldGrey = exports.dimGrey = exports.grey = exports.inverseWhite = exports.boldWhite = exports.dimWhite = exports.white = exports.chalk = void 0;
const chalk = require("chalk");
exports.chalk = chalk;
// The point of re-exporting is:
// 1. Fix typings to allow to pass `number` (very common case)
// 2. Easier/shorter to import, rather than from 'chalk'
// export type ColorFn = (...args: (string | number)[]) => string
exports.white = chalk.white;
exports.dimWhite = chalk.dim.white;
exports.boldWhite = chalk.bold.white;
exports.inverseWhite = chalk.inverse.white;
exports.grey = chalk.grey;
exports.dimGrey = chalk.dim.grey;
exports.boldGrey = chalk.bold.grey;
exports.yellow = chalk.yellow;
exports.dimYellow = chalk.dim.yellow;
exports.boldYellow = chalk.bold.yellow;
exports.inverseYellow = chalk.inverse.yellow;
exports.green = chalk.green;
exports.dimGreen = chalk.dim.green;
exports.boldGreen = chalk.bold.green;
exports.red = chalk.red;
exports.dimRed = chalk.dim.red;
exports.boldRed = chalk.bold.red;
exports.blue = chalk.blue;
exports.dimBlue = chalk.dim.blue;
exports.boldBlue = chalk.bold.blue;
exports.magenta = chalk.magenta;
exports.dimMagenta = chalk.dim.magenta;
exports.boldMagenta = chalk.bold.magenta;
exports.cyan = chalk.cyan;
exports.dimCyan = chalk.dim.cyan;
exports.boldCyan = chalk.bold.cyan;
