// @naturalcycles/nodejs-lib/dist/security/hash.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.base64ToBuffer = exports.bufferToBase64 = exports.base64ToString = exports.stringToBase64 = exports.hashAsBuffer = exports.hash = exports.md5AsBuffer = exports.md5 = void 0;
const crypto = require("crypto");
function md5(s) {
    return hash(s, 'md5');
}
exports.md5 = md5;
function md5AsBuffer(s) {
    return hashAsBuffer(s, 'md5');
}
exports.md5AsBuffer = md5AsBuffer;
function hash(s, algorithm) {
    return crypto.createHash(algorithm).update(s).digest('hex');
}
exports.hash = hash;
function hashAsBuffer(s, algorithm) {
    return crypto.createHash(algorithm).update(s).digest();
}
exports.hashAsBuffer = hashAsBuffer;
function stringToBase64(s) {
    return Buffer.from(s, 'utf8').toString('base64');
}
exports.stringToBase64 = stringToBase64;
function base64ToString(strBase64) {
    return Buffer.from(strBase64, 'base64').toString('utf8');
}
exports.base64ToString = base64ToString;
function bufferToBase64(b) {
    return b.toString('base64');
}
exports.bufferToBase64 = bufferToBase64;
function base64ToBuffer(strBase64) {
    return Buffer.from(strBase64, 'base64');
}
exports.base64ToBuffer = base64ToBuffer;
