// @naturalcycles/dev-lib/dist/testing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expectResults = exports.mockAllKindsOfThings = exports.timekeeper = exports.MOCK_TS_2018_06_21 = exports.resetTime = exports.mockTimeMillis = exports.mockTime = exports.silentConsole = exports.deepFreeze = exports.jestOffline = void 0;
const timekeeper = require("timekeeper");
exports.timekeeper = timekeeper;
const jestOffline_util_1 = require("../jestOffline.util");
Object.defineProperty(exports, "jestOffline", { enumerable: true, get: function () { return jestOffline_util_1.jestOffline; } });
const expect_util_1 = require("./expect.util");
Object.defineProperty(exports, "expectResults", { enumerable: true, get: function () { return expect_util_1.expectResults; } });
const mockAllKindsOfThings_1 = require("./mockAllKindsOfThings");
Object.defineProperty(exports, "mockAllKindsOfThings", { enumerable: true, get: function () { return mockAllKindsOfThings_1.mockAllKindsOfThings; } });
const testing_util_1 = require("./testing.util");
Object.defineProperty(exports, "deepFreeze", { enumerable: true, get: function () { return testing_util_1.deepFreeze; } });
Object.defineProperty(exports, "silentConsole", { enumerable: true, get: function () { return testing_util_1.silentConsole; } });
const time_util_1 = require("./time.util");
Object.defineProperty(exports, "mockTime", { enumerable: true, get: function () { return time_util_1.mockTime; } });
Object.defineProperty(exports, "mockTimeMillis", { enumerable: true, get: function () { return time_util_1.mockTimeMillis; } });
Object.defineProperty(exports, "MOCK_TS_2018_06_21", { enumerable: true, get: function () { return time_util_1.MOCK_TS_2018_06_21; } });
Object.defineProperty(exports, "resetTime", { enumerable: true, get: function () { return time_util_1.resetTime; } });
