// @naturalcycles/nodejs-lib/dist/bin/kpy.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const yargs = require("yargs");
const kpy_1 = require("../fs/kpy");
const script_1 = require("../script");
(0, script_1.runScript)(() => {
    const { _: [baseDir, ...inputPatterns], ...opt } = yargs.demandCommand(2).options({
        silent: {
            type: 'boolean',
            descr: 'Suppress all text output', // todo: desc!
        },
        verbose: {
            type: 'boolean',
            descr: 'Report progress on every file',
        },
        overwrite: {
            type: 'boolean',
            default: true,
        },
        dotfiles: {
            type: 'boolean',
        },
        flat: {
            type: 'boolean',
        },
        dry: {
            type: 'boolean',
        },
        move: {
            type: 'boolean',
            descr: 'Move files instead of copy',
        },
    }).argv;
    const outputDir = inputPatterns.pop();
    /*
    console.log({
      argv: process.argv,
      baseDir,
      inputPatterns,
      outputDir,
      silent,
      overwrite,
    })*/
    const kpyOpt = {
        baseDir: baseDir,
        inputPatterns: inputPatterns,
        outputDir,
        ...opt,
        noOverwrite: !opt.overwrite,
    };
    (0, kpy_1.kpySync)(kpyOpt);
});
